### PR TITLE
Add support for datagram packing

### PR DESCRIPTION
**Pull request by hanno-arm**
__Summary:__ This PR adds support for stacking multiple records in a single datagram, based on the ongoing work to support outgoing handshake fragmentation in #1939. The first new commit is https://github.com/ARMmbed/mbedtls/pull/1918/commits/5aa4e2cedd819b4fd307531eaa64f4f5ca8a01d4.

### CODE REVIEW

**Comment by hanno-arm** [comment]
This is on hold until the test failures in #1879 have been resolved.

**Comment by hanno-arm** [comment]
All CI interoperability failures on the underlying #1879 have been understood and seen as not caused by an issue in Mbed TLS. Testing this again.

**Comment by hanno-arm** [comment]
rest

**Comment by hanno-arm** [comment]
retest

**Comment by hanno-arm** [comment]
@mpg @AndrzejKurek Do you think you will have time to review this?

**Comment by mpg** [comment]
Yes, I'll review this next week!

**Comment by hanno-arm** [comment]
The introduction of datagram packing significantly changes the nature of the proxy tests in `ssl-opt.sh` since the UDP proxy works on the basis of datagrams: where it was previously capable of reordering or dropping some messages from flights, this is no longer possible if the entire flight is sent within a single datagram - as is the default behavior now. For example, without further changes to the proxy, or without the possibility to disable datagram packing, it is no longer possible to simulate a delayed CCS message. This is also the reason why we see the CI succeeding albeit it fails for the underlying fragmentation PR #1879.

I believe we should consider ways around this, for example
1. Allow static enabling/disabling of datagram packing at compile time.
2. Allow dynamic enabling/disabling of datagram packing at run time.
3. Improve the UDP proxy by having it dissect datagrams into records prior to performing its other operations.

Leaving tests aside, the only practical reason I see for allowing to disable datagram packing are potential interoperability issues with peers that happen to have problems receiving packed datagrams, but this seems quite unlikely.

Going through the options, the following are the consequences to the test system for each of them:
1. For static enabling/disabling of datagram packing, we'd need another build + test in `all.sh`. Some tests in `ssl-opt.sh` would need to be disabled for datagram packing (such as the one for delayed CCS's).
2. For dynamic enabling/disabling of datagram packing, this could be controlled through a new parameter to `ssl_client2` and `ssl_server2`, so that we could (a) disable datagram packing for the tests conflicting with it, and (b) duplicate tests which are meaningful with and without datagram packing.
3. When the UDP proxy not only supports packing but also splitting of datagrams, we'd need to turn that on in those tests that conflict with datagram packing, and duplicate the tests which are meaningful with and without datagram packing. 

@mpg What do you think?

**Comment by mpg** [comment]
@hanno-arm I'll reply to those question later because I have to go now, but I wanted to let you know I've put up a new PR with the fragmentation work here: https://github.com/ARMmbed/mbedtls/pull/1918 It has a cleaner history, fixes for a few bugs, more tests for MTU settings, and should pass CI (as it doesn't have the problematic interop tests yet).

I'd like to suggest you rebase your work on the new branch of mine, in order to have a cleaner history (currently there are a number of WIP/temporary/revert commits in the history) and the bug fixes.

**Comment by hanno-arm** [comment]
Thanks @mpg, I will do the rebase in the meantime.

**Comment by mpg** [comment]
> Leaving tests aside, the only practical reason I see for allowing to disable datagram packing are potential interoperability issues with peers that happen to have problems receiving packed datagrams, but this seems quite unlikely.

I'm not sure. IIRC, OpenSSL starts by sending packed datagrams, then if it has to retransmit tries unpacked, then it if has to retransmit again, it starts lowering the "estimated handshake MTU" and fragmenting if necessary. I don't think I've seen it stated anywhere, but I always assumed this was in order to enhance compatibility with peers that might not support datagram packing. However perhaps I'm misinterpreting, and even if I'm not, perhaps this is no longer a concern these days.

Anyway, I'm inclined to go for option 2 (dynamic setting of this behaviour) unless you think that will make the code too complex.

If we don't got for option 2, I quite dislike option 1 (static setting) as we have too many build options to test already, but before we go for option 3 I'd like to suggest option 4: find a peer that doesn't pack (I think you mentioned recently that `openssl s_server -debug` is such a peer) and use it for the tests that need it, so that we don't have to make the proxy smarter for now.

**Comment by mpg** [comment]
Aw, github is now showing commit in the wrong order. Apparently it does this [on purpose](https://help.github.com/articles/why-are-my-commits-in-the-wrong-order/) because it's "a space for discussion" - well I find it much easier to discuss a series of commits when it's presented in its logical order.

Anyway, reviewers beware of the github interface. Thankfully `git log --patch --reverse` is our reliable friend.

**Comment by mpg** [comment]
@hanno-arm Thanks for reworking the history! However I'm afraid you didn't quite rebase on top of #1939, but rather cherry-picked it or something. As a result, all the commits from #1939 are present with a different commit ID, which means when both PRs are merged, the history will contains "duplicate" commits with the same title and date but different hashes, which I tend to find pretty annoying when searching the history later on.

I fixed the issue by `reset-hard`ing my branch to `8c75a5d6a06f` then `force-push`ing it, so you don't have anything do to here, but I wanted to let you know.

In the future, tempted to base a branch on two other branches (that are also going to be merged independently), we should base on one and merge the other in, in order to avoid near-duplicate commits in the overall history.

**Comment by hanno-arm** [comment]
@mpg I based both the datagram packing and the fragmentation work on the fix for #1941, that's why the commits have different ID's.

**Comment by hanno-arm** [comment]
@mpg This is based on #1939 now.

**Comment by hanno-arm** [comment]
retest

**Comment by hanno-arm** [comment]
retest

**Comment by hanno-arm** [comment]
retest

**Comment by hanno-arm** [comment]
reset

**Comment by hanno-arm** [comment]
@AndrzejKurek @k-stachowiak I am currently debugging a CI failure in the `DTLS fragmenting: proxy MTU, resumed handshake` that I cannot reproduce locally, and am hence adding and force pushing some test changes to see if / how they affect the failure. Please ignore these UDP proxy related commits for now.

**Comment by hanno-arm** [comment]
retest

File: include/mbedtls/ssl.h [hunk 1]
```diff
@@ -1112,6 +1117,12 @@ struct mbedtls_ssl_context
     size_t out_msglen;          /*!< record header: message length    */
     size_t out_left;            /*!< amount of data not yet written   */
 
+    unsigned char cur_out_ctr[8]; /*!<  Outgoing record sequence  number. */
```
Comment (AndrzejKurek) [inline include/mbedtls/ssl.h]: Double space

File: library/ssl_cli.c [hunk 2]
```diff
@@ -1088,12 +1088,21 @@ static int ssl_write_client_hello( mbedtls_ssl_context *ssl )
...
 
-    if( ( ret = mbedtls_ssl_write_record( ssl ) ) != 0 )
+    if( ( ret = mbedtls_ssl_write_handshake_msg( ssl ) ) != 0 )
     {
-        MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_ssl_write_record", ret );
+        MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_ssl_write_handshake_msg", ret );
         return( ret );
     }
 
+#if defined(MBEDTLS_SSL_PROTO_DTLS)
```
Comment (AndrzejKurek) [inline library/ssl_cli.c]: This code is used once here, and twice in the server code - it could be extracted into a shared function.

File: library/ssl_tls.c [hunk 3]
```diff
@@ -96,7 +98,75 @@ static int ssl_check_timer( mbedtls_ssl_context *ssl )
...
+}
+
+static int ssl_get_remaining_space_in_datagram( mbedtls_ssl_context const *ssl )
+{
+    size_t   const bytes_written = ssl->out_left;
+    uint16_t const mtu           = ssl_get_maximum_datagram_size( ssl );
+
+    /* Double-check that the write-index hasn't gone
+     * past what we can transmit in a single datagram. */
+    if( bytes_written > (size_t) mtu )
```
Comment (AndrzejKurek) [inline library/ssl_tls.c]: Can this fail if mbedtls_ssl_set_mtu is called mid-connection?

File: library/ssl_cli.c (hunk 2 above)
Comment (hanno-arm) [reply library/ssl_cli.c]: I agree, and it should be subsumed with `mbedtls_ssl_send_flight_completed()`, because they always go in tandem - however, this is coming from #1939, and it is not unlikely that this code might go away with MPS anyway, so I'd be inclined to leave things as they are for now.

File: include/mbedtls/ssl.h (hunk 1 above)
Comment (hanno-arm) [reply include/mbedtls/ssl.h]: Thanks, will fix it.

File: library/ssl_tls.c [hunk 4]
```diff
@@ -96,7 +98,75 @@ static int ssl_check_timer( mbedtls_ssl_context *ssl )
...
+    }
+
+    return( (int) ( mtu - bytes_written ) );
+}
+
+static int ssl_get_remaining_payload_in_datagram( mbedtls_ssl_context const *ssl )
+{
+    int ret;
+    size_t remaining, expansion;
+    size_t max_len = MBEDTLS_SSL_MAX_CONTENT_LEN;
```
Comment (AndrzejKurek) [inline library/ssl_tls.c]: I would suggest renaming it to max_content_len or something indicating that it's about content, for readability.

File: library/ssl_tls.c [hunk 5]
```diff
@@ -2822,54 +2881,168 @@ static void ssl_swap_epochs( mbedtls_ssl_context *ssl )
...
 /*
  * Retransmit the current flight of messages.
+ */
+int mbedtls_ssl_resend( mbedtls_ssl_context *ssl )
+{
+    int ret = 0;
+
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> mbedtls_ssl_resend" ) );
+
+    ret = mbedtls_ssl_flight_transmit( ssl );
```
Comment (AndrzejKurek) [inline library/ssl_tls.c]: This line could be moved to the int ret = [...] line, but I guess that if you want to have debugs in the right place... cosmetic remark.

File: library/ssl_tls.c [hunk 6]
```diff
@@ -2822,54 +2881,168 @@ static void ssl_swap_epochs( mbedtls_ssl_context *ssl )
...
+    int ret;
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> mbedtls_ssl_flight_transmit" ) );
 
     if( ssl->handshake->retransmit_state != MBEDTLS_SSL_RETRANS_SENDING )
     {
-        MBEDTLS_SSL_DEBUG_MSG( 2, ( "initialise resending" ) );
+        MBEDTLS_SSL_DEBUG_MSG( 2, ( "initialise flight transmission" ) );
 
         ssl->handshake->cur_msg = ssl->handshake->flight;
+        ssl->handshake->cur_msg_p = ssl->handshake->flight->p + 12;
```
Comment (AndrzejKurek) [inline library/ssl_tls.c]: Why 12? If we take the opportunity to define ones and zeros as SSL_FORCE_FLUSH and such, documenting this magic number is a must.

File: library/ssl_tls.c [hunk 7]
```diff
@@ -2822,54 +2881,168 @@ static void ssl_swap_epochs( mbedtls_ssl_context *ssl )
...
+
+        uint8_t const force_flush = ssl->disable_datagram_packing == 1 ?
+            SSL_FORCE_FLUSH : SSL_DONT_FORCE_FLUSH;
 
         /* Swap epochs before sending Finished: we can't do it after
          * sending ChangeCipherSpec, in case write returns WANT_READ.
          * Must be done before copying, may change out_msg pointer */
-        if( cur->type == MBEDTLS_SSL_MSG_HANDSHAKE &&
-            cur->p[0] == MBEDTLS_SSL_HS_FINISHED )
+        if( is_finished && ssl->handshake->cur_msg_p == ( cur->p + 12 ) )
```
Comment (AndrzejKurek) [inline library/ssl_tls.c]: Why 12? If we take the opportunity to define ones and zeros as SSL_FORCE_FLUSH and such, documenting this magic number is a must.

File: library/ssl_tls.c [hunk 8]
```diff
@@ -2822,54 +2881,168 @@ static void ssl_swap_epochs( mbedtls_ssl_context *ssl )
...
+            ssl->out_msglen  = cur->len;
+            ssl->out_msgtype = cur->type;
+
+            /* Update position inside current message */
+            ssl->handshake->cur_msg_p += cur->len;
+        }
+        else
+        {
+            const unsigned char * const p = ssl->handshake->cur_msg_p;
+            const size_t hs_len = cur->len - 12;
```
Comment (AndrzejKurek) [inline library/ssl_tls.c]: 12, again, as above

File: library/ssl_tls.c [hunk 9]
```diff
@@ -2822,54 +2881,168 @@ static void ssl_swap_epochs( mbedtls_ssl_context *ssl )
...
+            ssl->out_msg[8] = ( ( frag_off       ) & 0xff );
+
+            ssl->out_msg[ 9] = ( ( cur_hs_frag_len >> 16 ) & 0xff );
+            ssl->out_msg[10] = ( ( cur_hs_frag_len >>  8 ) & 0xff );
+            ssl->out_msg[11] = ( ( cur_hs_frag_len       ) & 0xff );
+
+            MBEDTLS_SSL_DEBUG_BUF( 3, "handshake header", ssl->out_msg, 12 );
 
-        ssl->handshake->cur_msg = cur->next;
+            /* Copy the handshame message content and set records fields */
```
Comment (AndrzejKurek) [inline library/ssl_tls.c]: handsha**K**e *

File: library/ssl_tls.c [hunk 10]
```diff
@@ -2927,43 +3100,92 @@ void mbedtls_ssl_send_flight_completed( mbedtls_ssl_context *ssl )
...
+ * DTLS: except for HelloRequest, messages are only queued, and will only be
+ * actually sent when calling flight_transmit() or resend().
+ *
+ * Inputs:
+ *  - ssl->out_msglen: 4 + actual handshake message len
+ *      (4 is the size of handshake headers for TLS)
+ *  - ssl->out_msg[0]: the handshake type (ClientHello, ServerHello, etc)
+ *  - ssl->out_msg + 4: the handshake message body
+ *
+ * Outputs, ie state before passing to flight_append() or write_record():
```
Comment (AndrzejKurek) [inline library/ssl_tls.c]: i.e.

File: library/ssl_tls.c [hunk 11]
```diff
@@ -2927,43 +3100,92 @@ void mbedtls_ssl_send_flight_completed( mbedtls_ssl_context *ssl )
...
-    MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> write record" ) );
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> write handshake message" ) );
+
+    /*
+     * Sanity checks
+     */
+    if( ssl->out_msgtype != MBEDTLS_SSL_MSG_HANDSHAKE &&
+        ssl->out_msgtype != MBEDTLS_SSL_MSG_CHANGE_CIPHER_SPEC )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "should never happen" ) );
```
Comment (AndrzejKurek) [inline library/ssl_tls.c]: Please try to write something more descriptive in the error/debug message, as it does not help much (especially when multiple places have the same string). Same for other places in this PR.

File: library/ssl_tls.c [hunk 12]
```diff
@@ -3080,18 +3337,64 @@ int mbedtls_ssl_write_record( mbedtls_ssl_context *ssl )
...
+         * the remaining space in the datagram. */
+        if( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
+        {
+            ret = ssl_get_maximum_datagram_size( ssl );
+            if( ret < 0 )
+                return( ret );
+
+            if( protected_record_size > (size_t) ret )
+            {
+                /* Should never happen */
```
Comment (AndrzejKurek) [inline library/ssl_tls.c]: Please add a descriptive debug here; It'll help when looking for a root cause of an error.

File: library/ssl_tls.c [hunk 13]
```diff
@@ -3080,18 +3337,64 @@ int mbedtls_ssl_write_record( mbedtls_ssl_context *ssl )
...
+
+        ssl->out_left += protected_record_size;
+        ssl->out_hdr  += protected_record_size;
+        ssl_update_out_pointers( ssl, ssl->transform_out );
+
+        for( i = 8; i > ssl_ep_len( ssl ); i-- )
+            if( ++ssl->cur_out_ctr[i - 1] != 0 )
+                break;
+
+        /* The loop goes to its end iff the counter is wrapping */
```
Comment (AndrzejKurek) [inline library/ssl_tls.c]: if*

File: library/ssl_tls.c [hunk 14]
```diff
@@ -6841,26 +7230,48 @@ int mbedtls_ssl_get_record_expansion( const mbedtls_ssl_context *ssl )
...
     const mbedtls_ssl_transform *transform = ssl->transform_out;
+    unsigned block_size;
+
+    if( transform == NULL )
+        return( (int) mbedtls_ssl_hdr_len( ssl ) );
 
 #if defined(MBEDTLS_ZLIB_SUPPORT)
     if( ssl->session_out->compression != MBEDTLS_SSL_COMPRESS_NULL )
         return( MBEDTLS_ERR_SSL_FEATURE_UNAVAILABLE );
+    }
```
Comment (AndrzejKurek) [inline library/ssl_tls.c]: Isn't this a stray bracket?

File: library/ssl_tls.c [hunk 15]
```diff
@@ -6881,19 +7292,64 @@ size_t mbedtls_ssl_get_max_frag_len( const mbedtls_ssl_context *ssl )
...
+        }
+
+        if( max_len > mtu - overhead )
+            max_len = mtu - overhead;
+    }
+#endif
+
+#if !defined(MBEDTLS_SSL_MAX_FRAGMENT_LENGTH) &&        \
+    !defined(MBEDTLS_SSL_PROTO_DTLS)
+    ((void) ssl);
```
Comment (AndrzejKurek) [inline library/ssl_tls.c]: Missing spaces

File: programs/ssl/ssl_client2.c [hunk 16]
```diff
@@ -198,7 +200,11 @@ int main( void )
 #define USAGE_DTLS \
     "    dtls=%%d             default: 0 (TLS)\n"                           \
     "    hs_timeout=%%d-%%d    default: (library default: 1000-60000)\n"    \
-    "                        range of DTLS handshake timeouts in millisecs\n"
+    "                        range of DTLS handshake timeouts in millisecs\n" \
+    "    mtu=%%d              default: (library default: unlimited)\n"  \
```
Comment (AndrzejKurek) [inline programs/ssl/ssl_client2.c]: inconsistent spaces

File: programs/ssl/ssl_server2.c [hunk 17]
```diff
@@ -297,7 +299,11 @@ int main( void )
 #define USAGE_DTLS \
     "    dtls=%%d             default: 0 (TLS)\n"                           \
     "    hs_timeout=%%d-%%d    default: (library default: 1000-60000)\n"    \
-    "                        range of DTLS handshake timeouts in millisecs\n"
+    "                        range of DTLS handshake timeouts in millisecs\n" \
+    "    mtu=%%d              default: (library default: unlimited)\n"  \
```
Comment (AndrzejKurek) [inline programs/ssl/ssl_server2.c]: inconsistent spaces

File: programs/ssl/ssl_server2.c [hunk 18]
```diff
@@ -2165,6 +2193,7 @@ int main( int argc, char *argv[] )
     };
 #endif
 
+
```
Comment (AndrzejKurek) [inline programs/ssl/ssl_server2.c]: Unnecessary space

File: include/mbedtls/ssl.h [hunk 19]
```diff
@@ -1374,6 +1385,46 @@ void mbedtls_ssl_set_bio( mbedtls_ssl_context *ssl,
...
+/**
+ * \brief          Set the Maximum Tranport Unit (MTU).
+ *                 Special value: 0 means unset (no limit).
+ *                 This represents the maximum size of a datagram payload
+ *                 handled by the transport layer (usually UDP) as determined
+ *                 by the network link and stack. In practice, this controls
+ *                 the maximum size datagram the DTLS layer will pass to the
+ *                 \c f_send() callback set using \c mbedtls_ssl_set_bio().
+ *
+ * \note           This can be called at any point during the connection, for
```
Comment (AndrzejKurek) [inline include/mbedtls/ssl.h]: Do we have any tests calling it during the connection?

File: library/ssl_tls.c (hunk 9 above)
Comment (mpg) [reply library/ssl_tls.c]: I think this is part of #1939 and was fixed there in the meantime there. Please be aware that this PR is based on #1939, so it might be a good idea to review commit by commit, and comment here only on commits that were added in this PR (and you're obviously very welcome to comment on #1939 directly if you want to).

Also, you should be aware that [github is deliberately showing you commits in the wrong order](https://github.com/ARMmbed/mbedtls/pull/1918#issuecomment-412812630).

File: programs/ssl/ssl_client2.c (hunk 16 above)
Comment (mpg) [reply programs/ssl/ssl_client2.c]: I believe those spaces are actually consistent once the `%%` has been turned into a single `%`, that is when the usage message is actually printed out to the user.

File: library/ssl_tls.c (hunk 9 above)
Comment (AndrzejKurek) [reply library/ssl_tls.c]: That's the way I'd normally review it, but, given the time constraints, I'll rather review the full PR at one go.

Comment (hanno-arm) [reply library/ssl_tls.c]: @AndrzejKurek Could you nonetheless try to put the comments in the correct PR? I don't think it's good to start mixing the two PR's by having me address some issues from #1939.

File: library/ssl_tls.c (hunk 11 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: @AndrzejKurek I don't mean to argue that's how things be on the long run, but we're quite consistently using `should never happen` for internal assertion failures. Note also that you can tell them apart because the debug will include the file and line.

File: library/ssl_tls.c (hunk 12 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: Same as above.

File: library/ssl_tls.c (hunk 13 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: @AndrzejKurek That's intentional - `iff` is a common abbreviation for `if and only if`.

File: library/ssl_tls.c (hunk 15 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: I don't know the origin of this, but the use of `((void) xxx);` is consistent across the library.

File: include/mbedtls/ssl.h (hunk 19 above)
Comment (hanno-arm) [reply include/mbedtls/ssl.h]: This is for @mpg.

File: library/ssl_tls.c (hunk 10 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: This is for @mpg.

File: library/ssl_tls.c (hunk 5 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: I prefer it this way, leaving the debugging message signalling the function entry prior to what the function actually does.

File: library/ssl_tls.c (hunk 6 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: The `force_flush` field was newly introduced, while the use of `12` as a magic constant is pre-existing. I am generally reluctant to do too much cleanup in this PR because it is temporary and likely to go away soon anyway. If you insist, though, I will change it.

File: library/ssl_tls.c (hunk 14 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: Gosh, yes... don't know how that sneaked in.

File: library/ssl_tls.c (hunk 3 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: No, because the transmission of records is atomic from the perspective of the user - the only way the user could change the MTU during transmission would be to call `set_mtu` after receiving `WANT_WRITE`, which is an API violation.

**Comment by hanno-arm** [comment]
retest

File: library/ssl_tls.c (hunk 13 above)
Comment (AndrzejKurek) [reply library/ssl_tls.c]: Oh, didn't know that, sorry!

**Comment by hanno-arm** [comment]
@AndrzejKurek @k-stachowiak I merged #1939 and #1951 into this PR; the only change that had to be made for this is using the new function `ssl_get_current_mtu()` introduced in #1951 in place of `ssl->mtu`. Please re-review.

**Comment by hanno-arm** [comment]
@AndrzejKurek @k-stachowiak Pushed 11682cc to fix a build failure on Windows due to conversion warnings.

**Comment by sbutcher-arm** [comment]
CI is failing in the MS Visual Studio build phase with:
```
C:\builds\ws\mbed-tls-pr\library\ssl_tls.c(3366): warning C4267: '=' : conversion from 'size_t' to 'int', possible loss of data [C:\builds\ws\mbed-tls-pr\library\mbedtls.vcxproj]
```

**Comment by hanno-arm** [comment]
@sbutcher-arm Fixed in 554b0af.

File: library/ssl_tls.c [hunk 20]
```diff
@@ -3146,11 +3146,19 @@ int mbedtls_ssl_write_handshake_msg( mbedtls_ssl_context *ssl )
     /*
      * Sanity checks
      */
-    if( ssl->out_msgtype != MBEDTLS_SSL_MSG_HANDSHAKE &&
+    if( ssl->out_msgtype != MBEDTLS_SSL_MSG_HANDSHAKE          &&
```
Comment (AndrzejKurek) [inline library/ssl_tls.c]: Why such strange formatting?

Comment (hanno-arm) [reply library/ssl_tls.c]: Just taste, I usually prefer it to be aligned with the bracket in the next line.

Comment (AndrzejKurek) [reply library/ssl_tls.c]: It's very inconsistent here:
3149 - aligning to next bracket
3156 - aligning to last &&
3165 - no alignment.
But that's just a cosmetic change, so that's not a blocker for me.

**Comment by sbutcher-arm** [comment]
@k-stachowiak - Could you please review again?

File: library/ssl_tls.c [hunk 21]
```diff
@@ -96,7 +98,76 @@ static int ssl_check_timer( mbedtls_ssl_context *ssl )
...
+    if( bytes_written > mtu )
+    {
+        /* Should never happen... */
+        return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
+    }
+
+    return( (int) ( mtu - bytes_written ) );
+}
+
+static int ssl_get_remaining_payload_in_datagram( mbedtls_ssl_context const *ssl )
```
Comment (mpg) [inline library/ssl_tls.c]: As discussed on Skype, packing mutliple records whose length adds to much more than the (set or negotiated) MFL into the same datagram might not be a good idea (especially if the low MFL was negociated by the client) because the peer is likely to be unable to received datagrams much larger than (MFL + record overhead). We might want to restrict the size of the datagrams we send based on MFL even if the standard doesn't require us to do so.

File: library/ssl_tls.c [hunk 22]
```diff
@@ -101,6 +101,17 @@ static void ssl_update_out_pointers( mbedtls_ssl_context *ssl,
 static void ssl_update_in_pointers( mbedtls_ssl_context *ssl,
                                     mbedtls_ssl_transform *transform );
 #if defined(MBEDTLS_SSL_PROTO_DTLS)
+
+static uint16_t ssl_get_maximum_datagram_size( mbedtls_ssl_context const *ssl )
+{
+    uint16_t mtu = ssl->conf->mtu;
+
+    if( mtu != 0 && mtu < MBEDTLS_SSL_OUT_BUFFER_LEN )
+        return( (int) mtu );
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply library/ssl_tls.c]: @mpg @k-stachowiak I'm sorry for not reading the code properly initially, but by now it's fixed because the CI complained about it, too.

File: library/ssl_tls.c (hunk 21 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: @mpg I agree and will adapt the PR accordingly.

File: library/ssl_tls.c [hunk 23]
```diff
@@ -1345,14 +1425,6 @@ static int ssl_encrypt_buf( mbedtls_ssl_context *ssl )
     MBEDTLS_SSL_DEBUG_BUF( 4, "before encrypt: output payload",
                       ssl->out_msg, ssl->out_msglen );
 
-    if( ssl->out_msglen > MBEDTLS_SSL_OUT_CONTENT_LEN )
```
Comment (mpg) [inline library/ssl_tls.c]: You seem to be moving this check from `ssl_encrypt_buf()` where it applies to all records, to `ssl_write_handshake_msg()` where it now only applies to handshake messages. Can you justify that the check was redundant for all non-handshake messages (that is, unless I'm mistaken, alerts and application data)? Off the top of my head I'm not sure I can. If not, please insert back this check here (and leave the new one in `ssl_write_handshake_msg()` in addition).

File: library/ssl_tls.c (hunk 21 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: Fixed in f4b010e - please have a look.

File: include/mbedtls/ssl.h [hunk 24]
```diff
@@ -1098,6 +1098,11 @@ struct mbedtls_ssl_context
     int keep_current_message;   /*!< drop or reuse current message
                                      on next call to record layer? */
 
+#if defined(MBEDTLS_SSL_PROTO_DTLS)
+    uint8_t disable_datagram_packing;  /*!< Disable packing multiple records
```
Comment (mpg) [inline include/mbedtls/ssl.h]: This should go in section "Flags (bitfields)" at the end of the structure. We're wasting up to 4 bytes (on 32-bit machines) here.

File: include/mbedtls/ssl.h [hunk 25]
```diff
@@ -1757,6 +1813,38 @@ void mbedtls_ssl_conf_dtls_badmac_limit( mbedtls_ssl_config *conf, unsigned limi
...
+ *                 since messages belonging to a single datagram will not
+ *                 be reordered in transit, and so future message buffering
+ *                 or flight retransmission (if no buffering is used) as
+ *                 means to deal with reordering are needed less frequently.
+ *
+ * \note           Application datagrams are not affected by this option and
+ *                 are currently always sent in separate datagrams.
+ *
+ */
+void mbedtls_ssl_conf_datagram_packing( mbedtls_ssl_context *ssl,
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/ssl.h]: I think the setting should be in the SSL context because we might at some point consider turning off packing if the peer appears to have problems with it.

I will change the name of the function.

File: include/mbedtls/ssl.h (hunk 24 above)
Comment (hanno-arm) [reply include/mbedtls/ssl.h]: Thanks, I will move it.

Comment (hanno-arm) [reply include/mbedtls/ssl.h]: @mpg Sorry, on second look I see that this section is within the SSL configuration, but not the SSL context.

File: include/mbedtls/ssl.h (hunk 25 above)
[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/ssl.h]: Fixed in 1841b0a.

File: include/mbedtls/ssl.h [hunk 26]
```diff
@@ -1757,6 +1813,38 @@ void mbedtls_ssl_conf_dtls_badmac_limit( mbedtls_ssl_config *conf, unsigned limi
...
+ *
+ * \note           Allowing datagram packing reduces the network load since
+ *                 there's less overhead if multiple messages share the same
+ *                 datagram. Also, it increases the handshake efficiency
+ *                 since messages belonging to a single datagram will not
+ *                 be reordered in transit, and so future message buffering
+ *                 or flight retransmission (if no buffering is used) as
+ *                 means to deal with reordering are needed less frequently.
+ *
+ * \note           Application datagrams are not affected by this option and
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply include/mbedtls/ssl.h]: Yes, fixed in eb57008.

**Comment by hanno-arm** [comment]
@mpg I have pushed some commits to address your comments, and left a comment on your point in regards to the length check in `ssl_encrypt_buf()`.

File: library/ssl_tls.c (hunk 21 above)
Comment (hanno-arm) [reply library/ssl_tls.c]: @mpg c92b5c8 adds tests for how the MFL affects the effective MTU.

**Comment by hanno-arm** [comment]
@mpg Tests checking the MFL effectively restricting the MTU are currently failing, I am investigating.

**Comment by hanno-arm** [comment]
One of the tests I just added to `ssl-opt.sh` had wrong expectations, I removed it in 69ca0ad.

File: library/ssl_tls.c (hunk 23 above)
Comment (mpg) [reply library/ssl_tls.c]: I think you convinced me that there are enough other checks to make this one redundant, and thanks for pointing out if had become incorrect, I missed that. So, I'm not requesting any change here.

File: include/mbedtls/ssl.h (hunk 24 above)
Comment (mpg) [reply include/mbedtls/ssl.h]: Oops, my mistake, please disregard that comment then.

File: tests/ssl-opt.sh [hunk 27]
```diff
@@ -5020,6 +5020,10 @@ run_test    "DTLS fragmenting: server only (max_frag_len)" \
             -c "found fragmented DTLS handshake message" \
             -C "error"
 
+# With the MFL extension, the server has no way of forcing
```
Comment (mpg) [inline tests/ssl-opt.sh]: I think you just re-discovered our biggest issue with the MFL extension :) Fortunately the IETF has a replacement underway: https://tools.ietf.org/html/draft-ietf-tls-record-limit-03#section-3

**Comment by mpg** [comment]
Note: `all.sh -k -r -m` passed locally on the latest commit 69ca0ad

File: tests/ssl-opt.sh [hunk 28]
```diff
@@ -5459,35 +5515,31 @@ run_test    "DTLS fragmenting: gnutls server, DTLS 1.0" \
             -c "fragmenting handshake message" \
             -C "error"
 
-# gnutls-cli always tries IPv6 first, and doesn't fall back to IPv4 with DTLS
-requires_ipv6
 requires_config_enabled MBEDTLS_SSL_PROTO_DTLS
```
Comment (sbutcher-arm) [inline tests/ssl-opt.sh]: I think we need to have a comment here that explains and justifies the use of `--insecure` here and below.

Comment (hanno-arm) [reply tests/ssl-opt.sh]: @sbutcher-arm I don't mind changing this here, but the change was introduced by @mpg in #1939, and if your feedback on that PR leads to a 'fork' of #1939 being created anyway, I think it's better to add the change you requested here in that new PR, too.

Comment (sbutcher-arm) [reply tests/ssl-opt.sh]: Please liaise with @mpg and do what you think is best. I'd prefer it if you could do it, as I think you'll have more time than @mpg, but I'll leave it to you to decide. Just do it as a new PR, so I can get this one in without more rework.
